pkg/ride/environment.go

```diff
@@ -140,7 +325,225 @@ func (e *Environment) txID() rideType {
 }
 
 func (e *Environment) state() types.SmartState {
-	return e.st
+	return &e.st
+}
+
+func (e *Environment) setNewDAppAddress(address proto.Address) {
+	e.SetThisFromAddress(address)
+}
+
+func (e *Environment) applyToState(actions []proto.ScriptAction) error {
+
+	for _, action := range actions {
+		switch res := action.(type) {
+
+		case proto.DataEntryScriptAction:
+			if res.Entry.GetValueType() == proto.DataInteger {
+
+				bVal, err := res.Entry.MarshalValue()
+				if err != nil {
+					return errors.Wrap(err, "failed to marshal IntegerDataEntry bytes from value")
+				}
+				value := int64(binary.BigEndian.Uint64(bVal[1:]))
+
+				var intEntry proto.IntegerDataEntry
+				intEntry.Value = value
+				intEntry.Key = res.Entry.GetKey()
+
+				e.st.diff.dataEntries.diffInteger = append(e.st.diff.dataEntries.diffInteger, intEntry)
+			}
+
+			if res.Entry.GetValueType() == proto.DataBoolean {
+				bVal, err := res.Entry.MarshalValue()
+				if err != nil {
+					return errors.Wrap(err, "failed to marshal value value from Bool Entry")
+				}
+				value, err := proto.Bool(bVal[1:])
+				if err != nil {
+					return errors.Wrap(err, "failed to cast bytes value to bool format")
+				}
+
+				var boolEntry proto.BooleanDataEntry
+				boolEntry.Value = value
+				boolEntry.Key = res.Entry.GetKey()
+
+				e.st.diff.dataEntries.diffBool = append(e.st.diff.dataEntries.diffBool, boolEntry)
+			}
+
+			if res.Entry.GetValueType() == proto.DataBinary {
+				bVal, err := res.Entry.MarshalValue()
+				if err != nil {
+					return errors.Wrap(err, "failed to marshal value value from Binary Entry")
+				}
+				value, err := proto.BytesWithUInt16Len(bVal[1:])
+				if err != nil {
+					return errors.Wrap(err, "failed to cast binary value to binary format")
+				}
+
+				var binaryEntry proto.BinaryDataEntry
+				binaryEntry.Value = value
+				binaryEntry.Key = res.Entry.GetKey()
+
+				e.st.diff.dataEntries.diffBinary = append(e.st.diff.dataEntries.diffBinary, binaryEntry)
+			}
+
+			if res.Entry.GetValueType() == proto.DataString {
+				bVal, err := res.Entry.MarshalValue()
+				if err != nil {
+					return errors.Wrap(err, "failed to marshal value value from String Entry")
+				}
+				value, err := proto.StringWithUInt16Len(bVal[1:])
+				if err != nil {
+					return errors.Wrap(err, "failed to cast binary value to binary format")
+				}
+
+				var stringEntry proto.StringDataEntry
+				stringEntry.Value = value
+				stringEntry.Key = res.Entry.GetKey()
+
+				e.st.diff.dataEntries.diffString = append(e.st.diff.dataEntries.diffString, stringEntry)
+			}
+
+			if res.Entry.GetValueType() == proto.DataDelete {
+
+				key := res.Entry.GetKey()
+
+				for i, intDataEntry := range e.st.diff.dataEntries.diffInteger {
+					if key == intDataEntry.Key {
+						length := len(e.st.diff.dataEntries.diffInteger)
```

Comment:
No, no, no! Deleting doesn't mean to delete entry from local storage, but replace it with empty value. Removing from local storage just lead to retrieving it from state.